// eslint-plugin-testing-library/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const rules_1 = __importDefault(require("./rules"));
const domRules = {
    'testing-library/await-async-query': 'error',
    'testing-library/await-async-utils': 'error',
    'testing-library/no-await-sync-query': 'error',
    'testing-library/no-promise-in-fire-event': 'error',
    'testing-library/no-wait-for-empty-callback': 'error',
    'testing-library/prefer-find-by': 'error',
    'testing-library/prefer-screen-queries': 'error',
};
const angularRules = Object.assign(Object.assign({}, domRules), { 'testing-library/no-container': 'error', 'testing-library/no-debug': 'error', 'testing-library/no-dom-import': ['error', 'angular'], 'testing-library/no-node-access': 'error', 'testing-library/render-result-naming-convention': 'error' });
const reactRules = Object.assign(Object.assign({}, domRules), { 'testing-library/no-container': 'error', 'testing-library/no-debug': 'error', 'testing-library/no-dom-import': ['error', 'react'], 'testing-library/no-node-access': 'error', 'testing-library/render-result-naming-convention': 'error' });
const vueRules = Object.assign(Object.assign({}, domRules), { 'testing-library/await-fire-event': 'error', 'testing-library/no-container': 'error', 'testing-library/no-debug': 'error', 'testing-library/no-dom-import': ['error', 'vue'], 'testing-library/no-node-access': 'error', 'testing-library/render-result-naming-convention': 'error' });
module.exports = {
    rules: rules_1.default,
    configs: {
        dom: {
            plugins: ['testing-library'],
            rules: domRules,
        },
        angular: {
            plugins: ['testing-library'],
            rules: angularRules,
        },
        react: {
            plugins: ['testing-library'],
            rules: reactRules,
        },
        vue: {
            plugins: ['testing-library'],
            rules: vueRules,
        },
    },
};
