// eslint-plugin-testing-library/rules/await-async-query.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const node_utils_1 = require("../node-utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
exports.RULE_NAME = 'await-async-query';
exports.default = create_testing_library_rule_1.createTestingLibraryRule({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Enforce promises from async queries to be handled',
            category: 'Best Practices',
            recommended: 'warn',
        },
        messages: {
            awaitAsyncQuery: 'promise returned from {{ name }} query must be handled',
            asyncQueryWrapper: 'promise returned from {{ name }} wrapper over async query must be handled',
        },
        schema: [],
    },
    defaultOptions: [],
    create(context, _, helpers) {
        const functionWrappersNames = [];
        function detectAsyncQueryWrapper(node) {
            const innerFunction = node_utils_1.getInnermostReturningFunction(context, node);
            if (innerFunction) {
                functionWrappersNames.push(node_utils_1.getFunctionName(innerFunction));
            }
        }
        return {
            'CallExpression Identifier'(node) {
                if (helpers.isAsyncQuery(node)) {
                    detectAsyncQueryWrapper(node);
                    const closestCallExpressionNode = node_utils_1.findClosestCallExpressionNode(node, true);
                    if (!closestCallExpressionNode || !closestCallExpressionNode.parent) {
                        return;
                    }
                    const references = node_utils_1.getVariableReferences(context, closestCallExpressionNode.parent);
                    if (references && references.length === 0) {
                        if (!node_utils_1.isPromiseHandled(node)) {
                            return context.report({
                                node,
                                messageId: 'awaitAsyncQuery',
                                data: { name: node.name },
                            });
                        }
                    }
                    for (const reference of references) {
                        if (experimental_utils_1.ASTUtils.isIdentifier(reference.identifier) &&
                            !node_utils_1.isPromiseHandled(reference.identifier)) {
                            return context.report({
                                node,
                                messageId: 'awaitAsyncQuery',
                                data: { name: node.name },
                            });
                        }
                    }
                }
                else if (functionWrappersNames.includes(node.name)) {
                    if (!node_utils_1.isPromiseHandled(node)) {
                        return context.report({
                            node,
                            messageId: 'asyncQueryWrapper',
                            data: { name: node.name },
                        });
                    }
                }
            },
        };
    },
});
