// eslint-plugin-testing-library/rules/prefer-screen-queries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const experimental_utils_1 = require("@typescript-eslint/experimental-utils");
const node_utils_1 = require("../node-utils");
const create_testing_library_rule_1 = require("../create-testing-library-rule");
exports.RULE_NAME = 'prefer-screen-queries';
const ALLOWED_RENDER_PROPERTIES_FOR_DESTRUCTURING = [
    'container',
    'baseElement',
];
function usesContainerOrBaseElement(node) {
    const secondArgument = node.arguments[1];
    return (node_utils_1.isObjectExpression(secondArgument) &&
        secondArgument.properties.some((property) => node_utils_1.isProperty(property) &&
            experimental_utils_1.ASTUtils.isIdentifier(property.key) &&
            ALLOWED_RENDER_PROPERTIES_FOR_DESTRUCTURING.includes(property.key.name)));
}
exports.default = create_testing_library_rule_1.createTestingLibraryRule({
    name: exports.RULE_NAME,
    meta: {
        type: 'suggestion',
        docs: {
            description: 'Suggest using screen while querying',
            category: 'Best Practices',
            recommended: 'error',
        },
        messages: {
            preferScreenQueries: 'Avoid destructuring queries from `render` result, use `screen.{{ name }}` instead',
        },
        schema: [],
    },
    defaultOptions: [],
    create(context, _, helpers) {
        function reportInvalidUsage(node) {
            context.report({
                node,
                messageId: 'preferScreenQueries',
                data: {
                    name: node.name,
                },
            });
        }
        function saveSafeDestructuredQueries(node) {
            if (node_utils_1.isObjectPattern(node.id)) {
                for (const property of node.id.properties) {
                    if (node_utils_1.isProperty(property) &&
                        experimental_utils_1.ASTUtils.isIdentifier(property.key) &&
                        helpers.isBuiltInQuery(property.key)) {
                        safeDestructuredQueries.push(property.key.name);
                    }
                }
            }
        }
        const safeDestructuredQueries = [];
        const withinDeclaredVariables = [];
        return {
            VariableDeclarator(node) {
                if (!node_utils_1.isCallExpression(node.init) ||
                    !experimental_utils_1.ASTUtils.isIdentifier(node.init.callee)) {
                    return;
                }
                const isComingFromValidRender = helpers.isRenderUtil(node.init.callee);
                if (!isComingFromValidRender) {
                    saveSafeDestructuredQueries(node);
                }
                const isWithinFunction = node.init.callee.name === 'within';
                const usesRenderOptions = isComingFromValidRender && usesContainerOrBaseElement(node.init);
                if (!isWithinFunction && !usesRenderOptions) {
                    return;
                }
                if (node_utils_1.isObjectPattern(node.id)) {
                    saveSafeDestructuredQueries(node);
                    return;
                }
                if (experimental_utils_1.ASTUtils.isIdentifier(node.id)) {
                    withinDeclaredVariables.push(node.id.name);
                }
            },
            'CallExpression > Identifier'(node) {
                if (!helpers.isBuiltInQuery(node)) {
                    return;
                }
                if (!safeDestructuredQueries.some((queryName) => queryName === node.name)) {
                    reportInvalidUsage(node);
                }
            },
            'MemberExpression > Identifier'(node) {
                function isIdentifierAllowed(name) {
                    return ['screen', ...withinDeclaredVariables].includes(name);
                }
                if (!helpers.isBuiltInQuery(node)) {
                    return;
                }
                if (experimental_utils_1.ASTUtils.isIdentifier(node) &&
                    node_utils_1.isMemberExpression(node.parent) &&
                    node_utils_1.isCallExpression(node.parent.object) &&
                    experimental_utils_1.ASTUtils.isIdentifier(node.parent.object.callee) &&
                    node.parent.object.callee.name !== 'within' &&
                    helpers.isRenderUtil(node.parent.object.callee) &&
                    !usesContainerOrBaseElement(node.parent.object)) {
                    reportInvalidUsage(node);
                    return;
                }
                if (node_utils_1.isMemberExpression(node.parent) &&
                    experimental_utils_1.ASTUtils.isIdentifier(node.parent.object) &&
                    !isIdentifierAllowed(node.parent.object.name)) {
                    reportInvalidUsage(node);
                }
            },
        };
    },
});
